src/liballoc/rc.rs

```diff
@@ -1702,14 +1700,18 @@ impl<T> Weak<T> {
     /// This can be used to safely get a strong reference (by calling [`upgrade`]
     /// later) or to deallocate the weak count by dropping the `Weak<T>`.
     ///
-    /// It takes ownership of one weak count. In case a [`null`] is passed, a dangling [`Weak`] is
-    /// returned.
+    /// It takes ownership of one weak count (with the exception of pointers created from [`new`],
+    /// these don't have corresponding weak count).
     ///
     /// # Safety
     ///
-    /// The pointer must represent one valid weak count. In other words, it must point to `T` which
-    /// is or *was* managed by an [`Rc`] and the weak count of that [`Rc`] must not have reached
-    /// 0. It is allowed for the strong count to be 0.
+    /// The pointer must have originated from the [`from_raw`] (or [`as_raw`], provided there was
+    /// corresponding [`forget`] on the `Weak<T>`) and must still own its potential weak reference
```

Comment:
This isn't perfectly accurate, as you don't have to forget the original weak in the past, you can forget this weak if the pointer is from as_raw.

I'm not sure how to reword this note, though. Probably something along the lines of "only as many `Weak` are dropped as are created by non-raw methods".

Reply:
Technically yes, but only as long as some other `Weak<T>` or `Strong<T>` pointing to the same thing exists. If you created it after the weak count dropped to 0 and then forgot it, it would be UB, because it would play with the weak count field that no longer exists.

But if you hold a `Weak<T>` somewhere, then you probably don't really need to create and forget it. So I believe this might be good enough ‒ especially because this safety section errs on the safe side ‒ it's more restrictive than needs to be.

Anyway, I don't see how to make it more accurate, still readable and not ramble about it over two pages :-(.